### PR TITLE
Make parallel build memory threshold configurable

### PR DESCRIPTION
**Pull request by saschagrunert**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

The amount of memory required to build binaries in parallel is right now
set to 40GiB. We now make this variable to be able to build artifacts in
parallel even with a lower amount of memory.

This enables SIG Release to speed-up the build time drastically in
Google Cloud Build (GCB).

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

cc @kubernetes/release-engineering 
/sig release

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
None
```

### CODE REVIEW

**Review by cpanato — APPROVED** [review]
/lgtm

**Review by xmudrii — APPROVED** [review]
/lgtm

**Review by hasheddan — APPROVED** [review]
/lgtm

**Comment by saschagrunert** [comment]
@kubernetes/milestone-maintainers can we get this over the finishing line?

**Comment by ehashman** [comment]
@saschagrunert that's not the issue, you're missing an approver: https://github.com/kubernetes/kubernetes/blob/6d43e2b3bb55237fbb64118dc13beac259745983/hack/OWNERS#L18-L31

go chase one of these fine folks down :)

**Comment by cblecker** [comment]
/hold

So this variable is designed to be used as a deciding factor where if you have more than this amount of memory, you can build in parallel.. and below it, we don't because you experience errors during the build. I'm not clear on how making this configurable changes this.

I'd be okay with is making it configurable in general, but can we expand the comment to provide more context on what this variable does and why?

File: hack/lib/golang.sh
Comment (cblecker) [inline hack/lib/golang.sh:309]: This comment doesn't make sense anymore if this isn't hard coded

Comment (saschagrunert) [reply hack/lib/golang.sh]: Added a note that it can be overwritten at our own risk, whereas it still defaults to 40GB.

**Comment by saschagrunert** [comment]
@cblecker do you think you can give this another look? :)

**Comment by saschagrunert** [comment]
PING @fejta @deads2k @lavalamp may I ask you for an approval?

**Comment by BenTheElder** [comment]
If we're able to lower this, perhaps the default is higher than it needs to be?
I'd expect this to change over time with go releases, kubernetes changes ...

**Comment by saschagrunert** [comment]
> If we're able to lower this, perhaps the default is higher than it needs to be?
> I'd expect this to change over time with go releases, kubernetes changes ...

Yeah probably, this was just a test coming from our GCB instance. I was just a bit conservative in changing the default because I did not really know which impact it would have on existing build systems using older go versions. WDYT?

**Comment by BenTheElder** [comment]
> Yeah probably, this was just a test coming from our GCB instance. I was just a bit conservative in changing the default because I did not really know which impact it would have on existing build systems using older go versions. WDYT?

Your GCB instance is using newer than k/k? Or is the concern about backporting to older branches?

Ideally we should have a reasonable default on the current branch, which will evolve with time?

**Comment by saschagrunert** [comment]
> > Yeah probably, this was just a test coming from our GCB instance. I was just a bit conservative in changing the default because I did not really know which impact it would have on existing build systems using older go versions. WDYT?
> 
> Your GCB instance is using newer than k/k? Or is the concern about backporting to older branches?

Yeah it was mainly about the backports.

> Ideally we should have a reasonable default on the current branch, which will evolve with time?

I think we could play around with the value together with CI signal as follow-up.

**Comment by saschagrunert** [comment]
@BenTheElder do you think we can move forward with this PR and benchmark the actual required threshold after we made iit configurable?

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/hold cancel

**Comment by BenTheElder** [comment]
Sorry E_TOO_MANY_PRS. Looks like Aaron got it 😅

**Comment by spiffxp** [comment]
/test pull-kubernetes-unit
prow's status-reconciler should have triggered this but didn't, not sure why

**Pull request merged**
